components/Banner/examples/supplemental/MDCBannerViewLayout.m

```diff
@@ -0,0 +1,283 @@
+        currentXAxis += self.imageContainerFrame.size.width;
+        currentXAxis += kSpaceBetweenIconImageAndTextLabel;
+      }
+      CGSize originalTextLabelSize = self.textLabelFrame.size;
+      self.textLabel.frame = CGRectMake(currentXAxis, currentYAxis, originalTextLabelSize.width,
+                                        originalTextLabelSize.height);
+      currentXAxis = self.frameSize.width;
+      currentYAxis = self.frameSize.height;
+      currentXAxis -= kTrailingPadding;
+      currentYAxis -= kBottomPadding;
+      NSUInteger buttonCount = self.internalButtonFrames.count;
+      for (NSUInteger index = 0; index <= buttonCount - 1; ++index) {
+        CGRect buttonFrame = [self.internalButtonFrames[buttonCount - index - 1] CGRectValue];
+        currentYAxis -= buttonFrame.size.height;
+        CGRect newButtonFrame = CGRectMake(currentXAxis - buttonFrame.size.width, currentYAxis,
+                                           buttonFrame.size.width, buttonFrame.size.height);
+        self.internalButtonFrames[buttonCount - index - 1] =
+            [NSValue valueWithCGRect:newButtonFrame];
+        currentYAxis -= kButtonVerticalIntervalSpace;
+      }
+      break;
+    }
+  }
+}
+
+#pragma mark - Internal Layout Value Helpers
+
+- (CGFloat)buttonsWidthSum {
+  CGFloat buttonsWidthSum = 0.0f;
+  for (NSValue *buttonFrame in self.internalButtonFrames) {
+    buttonsWidthSum += [buttonFrame CGRectValue].size.width;
+  }
+  if (self.internalButtonFrames.count > 0) {
+    buttonsWidthSum += (self.internalButtonFrames.count - 1) * kButtonHorizontalIntervalSpace;
+  }
+  return buttonsWidthSum;
+}
+
+- (CGFloat)buttonsHeightSum {
+  CGFloat buttonsHeightSum = 0.0f;
+  for (NSValue *buttonFrame in self.internalButtonFrames) {
+    buttonsHeightSum += [buttonFrame CGRectValue].size.height;
+  }
+  if (self.internalButtonFrames.count > 0) {
+    buttonsHeightSum += (self.internalButtonFrames.count - 1) * kButtonVerticalIntervalSpace;
+  }
+  return buttonsHeightSum;
+}
+
+- (BOOL)isAbleToFitTextLabelWithWidthLimit:(CGFloat)widthLimit {
+  CGSize size =
+      [self.textLabel.text sizeWithAttributes:@{NSFontAttributeName : self.textLabel.font}];
+  return size.width <= widthLimit;
+}
+
+- (CGFloat)maximumHeightAmongViews:(NSArray<__kindof UIView *> *)views {
+  CGFloat maximumHeight = 0.0f;
+  for (UIView *view in views) {
+    maximumHeight = MAX(maximumHeight, view.frame.size.height);
+  }
+  return maximumHeight;
+}
+
+- (CGFloat)maximumButtonHeight {
+  CGFloat maximumHeight = 0.0f;
+  for (NSValue *buttonFrame in self.internalButtonFrames) {
+    maximumHeight = MAX(maximumHeight, [buttonFrame CGRectValue].size.height);
+  }
+  return maximumHeight;
+}
+
+#pragma mark - APIs
+
+- (NSArray *)buttonFrames {
+  return [self.internalButtonFrames copy];
+}
+
+- (CGRect)imageContainerFrame {
+  return self.imageContainer.frame;
+}
+
+- (CGRect)textLabelFrame {
+  return self.textLabel.frame;
+}
+
+@end
```

Comment:
There is a lot of interwoven logic in this file. Generally speaking at a high level I'd like to see:

- A method that can determine what layout style to use based on a given size.
- A method that, given a layout style and size, will return a fitted size.
- A method that, given a layout style, will commit the layout to the relevant views.

As it stands, I think these three different methods may all be woven together in updateLayoutStyleAndFrameSize.

Reply:
Addressed in 2b23608